src/main/scala/workflow/NodeOptimizationRule.scala

```diff
@@ -5,96 +5,341 @@ import org.apache.spark.rdd.RDD
+        // into the optimized instructions
+        val pipeToSplice = ot.optimize(inputs.head.asInstanceOf[RDD[a]], numPerPartition)
+        val instructionsToSplice = WorkflowUtils.pipelineToInstructions(pipeToSplice)
+
+        // First: Disconnect all the dependencies in the optimized pipeline on this TransformerApplyNode,
+        // and replace them with the spliceEndpoint
+        val spliceEndpoint = Pipeline.SOURCE - 1
+        val indexInNewPipeline = os.unoptimizedInstructionIndexToOptimizedIndex(index)
+        val removeResult = WorkflowUtils.disconnectAndRemoveInstructions(
+          Map(indexInNewPipeline -> spliceEndpoint),
+          os.optimizedInstructions)
+        val intermediateOptimizedInstructions = removeResult._1
+        val intermediateIndexMapping = os.unoptimizedInstructionIndexToOptimizedIndex andThen removeResult._2
+
+        // Then: Splice in the transformer's optimization into the optimized pipeline
+        // Note: We do not delete the optimizable transformer because it may be used elsewhere
+        // if we've merged equivalent TransformerNodes.
+        val spliceResult = WorkflowUtils.spliceInstructions(
+          instructionsToSplice,
+          intermediateOptimizedInstructions,
+          Map(Pipeline.SOURCE -> intermediateIndexMapping(inputIndices.head)),
+          spliceEndpoint)
+
+        os.copy(
+          optimizedInstructions = spliceResult._1,
+          unoptimizedInstructionIndexToOptimizedIndex = intermediateIndexMapping andThen spliceResult._2
+        )
+      }
+      case _: TransformerOutput => os
+      case _ => throw new ClassCastException("TransformerApplyNode dep wasn't pointing at a transformer")
+    }
+  }
+
+  /**
+   * Optimizes the estimator being fit by the instruction at the given index.
+   * Outputs a new optimization state.
+   *
+   * @param os The previous optimization state
+   * @param estimatorFitNode
+   * @param instructions
+   * @param index
+   * @return
+   */
+  private def optimizeEstimator(
+      os: OptimizationState,
+      estimatorFitNode: EstimatorFitNode,
+      instructions: Seq[Instruction],
+      index: Int
+    ): OptimizationState = {
+    val inputIndices = estimatorFitNode.inputs
+    val estIndex = estimatorFitNode.est
+
+    val numPerPartition = os.numPerPartitionPerNode(inputIndices.head)
+    val inputs = inputIndices.map(os.registers).collect {
+      case RDDOutput(rdd) => rdd
+    }
+
+    // Optimize the estimator if possible
+    os.registers(estIndex) match {
+      case EstimatorOutput(oe: OptimizableEstimator[a, b]) => {
+        val spliceEndpoint = Pipeline.SOURCE - 1
+
+        // Produce the optimized estimator fit instructions that we then
+        // need to splice into the pipeline
+        val dataSample = inputs.head.asInstanceOf[RDD[a]]
+        val pipeToSplice = oe.optimize(dataSample, numPerPartition).apply(dataSample)
```

Comment:
So what is happening here is that we're constructing a pipeline but w/ the sampled RDD instead of the true RDD, then we're removing the uses of the sampled RDD to produce instructions for a partial pipeline, that is the optimization of the `estimator`